### PR TITLE
feat(test-utils): add 'overview' function

### LINKED ISSUE #1461: Add a debuger function to check wrapper content
### What problem does this feature solve?
Actually, if you try to see the contents of your wrapper the result is a little bit confusing.
It could work when wrappers from simple elements, but if biger it tursn unreadable:

```javascript
console.log(simpleWrapper)
// output:  Wrapper { selector: 'p' }

console.log(bigerWrapper)
// output:  VueWrapper { isFunctionalComponent: undefined, emitted: [Object: null prototype] {},_emittedByOrder: [] }
```

One posible alternative could be to use `.html()` before log the content, but this function doesen't work with array wrappers.

My proposal is to add a function to the wrappers that allows to visualize its content in a clear and simple way, without worrying about what kind of wrapper it is. A well formated pretty print of the html of the wrapper element, who gives support to array wrappers too.

It can be very useful to understand what is really being tested, especially for people in their first steps who are still learning how the library works.

#### For simple wrapper:

```javascript
wrapper.debug()

// Output:
 Wrapper:
    <div>
      <p>1</p>
    </div> 
```

#### For array wrapper:

```javascript
wrapperArray.debug()

// Output:
Wrapper Array:
   At 0:
    <div>
      <p>1</p>
    </div> 

  At 1:
    <div>
      <p>2</p>
    </div> 

  At 2:
    <div>
      <p>3</p>
    </div> 
```

### PR DESCRIPTION
**Pull request by SergiCL**
close #1461

**What kind of change does this PR introduce?** (check at least one)

- [x] Feature

**Does this PR introduce a breaking change?** (check one)

- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch.
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue-test-utils/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:

- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
The main objective of this feature is "To give an overview of the state of the wrapper at the time the function is executed"

Example execution:
```javascript
wraper.overview()

// Output
Wrapper (Visible):

Html:
    <div class="test">
      <p>My name is Tess Ting</p>
    </div>

Data: {
    firstName: Tess,
    lastName: Ting
}

Computed: {
    fullName: Tess Ting'
}

Emitted: {',
    foo: [',
        0: [ hello, world ],
        1: [ bye, world ]'
    ],
    bar: [
        0: [ hey ]'
    ]
}
```

[More info](https://github.com/vuejs/vue-test-utils/issues/1461#issuecomment-602072301)

### CODE REVIEW

**Comment by lmiller1990** [comment]
Holy moly what a PR! I like how in depth your tests go.

Please give me some time to review this.

**Comment by SergiCL** [comment]
> Holy moly what a PR! I like how in depth your tests go.
>
> Please give me some time to review this.

Thank you. No problem, take as much time as you need.

File: test/specs/wrapper/overview.spec.js [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+          ''
+        ]
+
+        wrapper.isVisible = () => true
+        wrapper.overview()
+        expect(consoleOutput).to.have.ordered.members(expectedConsoleOutput)
+      })
+
+      it('does not print child component data or computed', () => {
+        if (mountingMethod.name === 'shallowMount') return
```
Comment (lmiller1990) [inline test/specs/wrapper/overview.spec.js]: I might be missing something - why are we skipping `shallowMount`?

File: test/specs/wrapper/overview.spec.js [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+          '',
+          'Html:',
+          '    <div class="test"></div>',
+          '',
+          'Data: {}',
+          '',
+          'Computed: {}',
+          '',
+          'Emitted: {',
+          '    foo: [',
```
Comment (lmiller1990) [inline test/specs/wrapper/overview.spec.js:211]: what is displayed if objects are emitted? Probably need a test for this

File: test/specs/wrapper/overview.spec.js [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    const nonVueWrapper = wrapper.find('p')
+    const message =
+      '[vue-test-utils]: wrapper.overview() can only be called on a Vue instance'
+
+    expect(() => nonVueWrapper.overview())
+      .to.throw()
+      .with.property('message', message)
+  })
+
+  if (vueVersion > 2) {
```
Comment (lmiller1990) [inline test/specs/wrapper/overview.spec.js:26]: how come we need this check? This lib only supports Vue 2.x - doe s something weird happen if you remove this?

**Review by lmiller1990 — COMMENTED** [review]
I left some comments!

File: test/specs/wrapper/overview.spec.js (hunk 1 above)
Comment (SergiCL) [reply test/specs/wrapper/overview.spec.js]: The test checks if the HTML shown in the overview includes the child component. 

With mount child component is:
`<div class="tester">test</div>`

But with shallow mount is something like this:
`<tester-stub></tester-stub>`

I don't know if it's really necessary to test this with shallow mount, but to make the test clearer I can add a diferent expect for every mounting method. 

For example:

```javascript
const expectedConsoleOutput = [
	  '',
	  'Wrapper (Visible):',
	  '',
	  'Html:',
	  mountingMethod.name === 'shallowMount'
	    ? '    <div>1<tester-stub></tester-stub>'
	    : '    <div>1<div class="tester">test</div>',
	  '    </div>',
	  '',
	  'Data: {}',
	  '',
	  'Computed: {}',
	  '',
	  'Emitted: {}',
	  ''
	]
```

Comment (SergiCL) [reply test/specs/wrapper/overview.spec.js]: > it ('does not print child component data or computed')

I got the test mixed up. You're talking about the one for computed and data. In this case it makes more sense, anyways the solution would be the same.

File: test/specs/wrapper/overview.spec.js (hunk 2 above)
Comment (SergiCL) [reply test/specs/wrapper/overview.spec.js:211]: I totally agree. I'll get right on it.

Edit: I will add a test to check when the object is emitted, and when the emit is empty.

File: test/specs/wrapper/overview.spec.js (hunk 3 above)
Comment (SergiCL) [reply test/specs/wrapper/overview.spec.js:26]: The test of `test:compat` fails at Vue 2.0.
This is because hooks are added to the emitted:

```javascript
"Emitted: {"
	" hook:beforeCreate: ["
	" 0: [ ]"
	" ],"
	" hook:created: ["
	" 0: [ ]"
	" ],"
	" hook:beforeMount: ["
	" 0: [ ]"
	" ],"
	" hook:mounted: ["
	" 0: [ ]"
	" ]"
 "}"
```

I don't know exactly how many people use version 2.0. If we want to test it, I can think of the following options:

1. **Duplicate the tests (or a part of them) in an `else`**: We repeat a lot of code but gain in readability.

2.  **Extract the expectedConsoleOutput to a function (at least the emitted)**: The tests are less clear, but it's better than adding a condition to each test.

Comment (SergiCL) [reply test/specs/wrapper/overview.spec.js:26]: I think I have a middle ground solution. 
Right now we're checking the outcome with this:

```javascript
expect(consoleOutput).to.have.ordered.members(expectedConsoleOutput)
```

If we replace the `have.ordered` with an `include`  it should work for Vue 2.0:
```
expect(consoleOutput).to.include.members(expectedConsoleOutput)
```

The only problem we have is when nothing has been emitted. In that case the expected emit will be `Emitted: {}` and in 2.0 it will show:

```javascript
"Emitted: {"
" hook:beforeCreate: ["
" 0: [ ]"
" ],"
" hook:created: ["
" 0: [ ]"
" ],"
" hook:beforeMount: ["
" 0: [ ]"
" ],"
" hook:mounted: ["
" 0: [ ]"
" ]"
"}"
```

My proposal is to create a simple function that takes care of this without making the tests dirty.

Something like:
```javascript
expect(compareConsoleOutput(consoleOutput, expected)).to.be.true
```
or even a function that includes the expect inside:
```javascript
expectToHaveConsoleOutputs(consoleOutput, expectedConsoleOutputs)
```

File: test/specs/wrapper/overview.spec.js (hunk 2 above)
Comment (lmiller1990) [reply test/specs/wrapper/overview.spec.js:211]: sounds good

File: test/specs/wrapper/overview.spec.js (hunk 1 above)
Comment (lmiller1990) [reply test/specs/wrapper/overview.spec.js]: I think it's fine to show the stubbed html for the shallow mounted components. shallowMount's output is basically useless, but I think it's better to be consistent and show it anyway. At least from my point of view, I would expect overview to show the same information for both rendering APIs.

File: test/specs/wrapper/overview.spec.js (hunk 3 above)
Comment (lmiller1990) [reply test/specs/wrapper/overview.spec.js:26]: I see - I think it's fine to leave the check in. I doubt many people are on 2.0 now anyway. No problem 👍

**Review by lmiller1990 — APPROVED** [review]
looks like a solid PR.

this is the last API we will be adding before v1! This will go out with v1, should be in a week or two.

**Comment by julisch94** [comment]
Hey folks, I'm unsure if this is the right place to ask. I would like such an `overview()` function but I can't find it anywhere even though I can see it was added by you. 

It's still in the code:
https://github.com/vuejs/vue-test-utils/blob/dev/packages/test-utils/src/wrapper.js#L450-L455C21
(even though deprecated, why?)

There's also nothing about it in the docs:
https://v1.test-utils.vuejs.org/api/wrapper

So how can I print the DOM tree of a component with vue test utils like I can do with react testing library using `screen.debug()`?

Anyone can help me understand this a bit better?

**Comment by lmiller1990** [comment]
You can print the HTML with `wrapper.html()` https://test-utils.vuejs.org/api/#html

I am not sure about `overview()`, I completely forgot about it. I think the confusion is the `overview()` function was added to v1 of this lib (which works with Vue 2) but you might be on v2 of this lib (which works with Vue 3).

**Comment by julisch94** [comment]
@lmiller1990 oh right! Thanks! I was looking at the v1 of the docs. Didn't notice that Google sent me to v1 instead of v2. And also thanks for pointing me to `wrapper.html()`. It served my purpose. :)
